### DISCUSSION TITLE
Keep getting "failed to stream in to volume" or when enabling P2P "failed to stream p2p out of volume" - 7.3.2

### DISCUSSION THREAD

**Discussion by skabba** [concourse/concourse#7348] [Help & Support]
I keep getting below errors when I upgraded from 6.4.0 to 7.3.2 this week:
<img width="737" alt="failed-to-stream" src="https://user-images.githubusercontent.com/24790166/127338926-111c33cb-5c1d-4947-b96b-7ce364633872.png">
I tried using P2P option, but then I get a similar error:
<img width="806" alt="failed-to-stream-p2p" src="https://user-images.githubusercontent.com/24790166/127338879-9f716940-2946-4053-8fc0-ca6054e7499b.png">

The annoying thing is that there is not really a pattern. Sometimes the same workers succeed, but mostly not.
Tried looking up the errors.

ATC:
```
Jul 28 13:37:35 ip-10-228-22-132.pnls.local web.sh[7617]: {"timestamp":"2021-07-28T13:37:35.513019239Z","level":"info","source":"atc","message":"atc.tracker.notify.run.task-step.stream-to.end","data":{"build":"59","build_id":1318843,"job":"update-pipeline
-flow-operatororder","job-id":27951,"pipeline":"pipeline-manager","session":"23.505.2.4.10","step-name":"generate-flow-operatororder-pipeline","team":"esb"}}
Jul 28 13:37:35 ip-10-228-22-132.pnls.local web.sh[7617]: {"timestamp":"2021-07-28T13:37:35.513071226Z","level":"error","source":"atc","message":"atc.tracker.notify.run.task-step.failed-to-stream-image-artifact-source","data":{"build":"59","build_id":1318
843,"container":"94e766d0-e1dd-4eb9-6b72-bcf5be6ffdc7","error":"failed to stream p2p out from volume","job":"update-pipeline-flow-operatororder","job-id":27951,"pipeline":"pipeline-manager","session":"23.505.2.4","step-name":"generate-flow-operatororder-p
ipeline","team":"esb"}}
Jul 28 13:37:35 ip-10-228-22-132.pnls.local web.sh[7617]: {"timestamp":"2021-07-28T13:37:35.704539615Z","level":"error","source":"atc","message":"atc.tracker.notify.run.task-step.failed-to-fetch-image-for-container","data":{"build":"59","build_id":1318843
,"container":"94e766d0-e1dd-4eb9-6b72-bcf5be6ffdc7","error":"failed to stream p2p out from volume","job":"update-pipeline-flow-operatororder","job-id":27951,"pipeline":"pipeline-manager","session":"23.505.2.4","step-name":"generate-flow-operatororder-pipe
line","team":"esb"}}
Jul 28 13:37:35 ip-10-228-22-132.pnls.local web.sh[7617]: {"timestamp":"2021-07-28T13:37:35.704610004Z","level":"info","source":"atc","message":"atc.tracker.notify.run.errored","data":{"build":"59","build_id":1318843,"error":"find or create container on w
orker ip-10-228-22-131.pnls.local: failed to stream p2p out from volume","job":"update-pipeline-flow-operatororder","pipeline":"pipeline-manager","session":"23.505.2","team":"esb"}}
```

On the worker nothing really specific like in the ATC logs. But do see this:
```
Jul 28 13:37:35 ip-10-228-22-131.pnls.local worker.sh[3068]: {"timestamp":"2021-07-28T13:37:35.520149426Z","level":"debug","source":"worker","message":"worker.container-sweeper.tick.run.command-exited","data":{"command":"sweep-containers","session":"6.75.
4"}}
...
Jul 28 13:37:35 ip-10-228-22-131.pnls.local worker.sh[3068]: {"timestamp":"2021-07-28T13:37:35.647538109Z","level":"info","source":"worker","message":"worker.garden.garden-server.run.spawned","data":{"handle":"8fee3319-0326-4277-57b1-8d36bac655c0","id":"a
1cd45b6-3016-46f7-7e58-6c1632e61234","session":"1.2.7437","spec":{"Path":"/opt/resource/check","Dir":"","User":"root","Limits":{},"TTY":null}}}
...
Jul 28 13:37:35 ip-10-228-22-131.pnls.local worker.sh[3068]: {"timestamp":"2021-07-28T13:37:35.868971116Z","level":"debug","source":"worker","message":"worker.volume-sweeper.tick.run.command-exited","data":{"command":"report-volumes","session":"7.75.3"}}
...
Jul 28 13:37:38 ip-10-228-22-131.pnls.local worker.sh[3068]: {"timestamp":"2021-07-28T13:37:38.118623170Z","level":"info","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.get-volume.volume-not-found","data":{"session":"5.1.14139.1","volume":"4b1868e9-2d59-4aba-66d4-648d6eedaded"}}
```

Here is my current setup (only showing settings I think do matter):

WEB:
```
CONCOURSE_SECRET_CACHE_ENABLED=true
CONCOURSE_SECRET_CACHE_DURATION=1440m
CONCOURSE_SECRET_CACHE_DURATION_NOTFOUND=1440m
CONCOURSE_ENABLE_GLOBAL_RESOURCES=true
CONCOURSE_ENABLE_P2P_VOLUME_STREAMING=true
```

WORKER:
```
CONCOURSE_BAGGAGECLAIM_BIND_IP=0.0.0.0
CONCOURSE_RUNTIME=containerd
```

Ports are exposed and open in the security group.

**Comment by evanchaoli** [comment]
Did you upgrade worker to 7.3 as well? When you `fly workers`, worker version should be 2.3.

**Reply by skabba** [reply]
Yes i did also upgrade all workers, created new EC2 instances with the new 7.3.2 binary on it. Fly was indeed showing 2.3.

**Reply by evanchaoli** [reply]
How many NICs does your worker have? What are those NIC names? Is there a load balancer in front of your workers?

**Reply by skabba** [reply]
1 NIC named eth0. And a lot of gdn virtual nics of course.
No load balancer in front of the workers. They use CONCOURSE_TSA_HOST.
Only an ALB in front of the WEB.

In the meantime i did a downgrade to 7.2.0, which for me has the same issues. Then did a rollback to 6.4.0 and no more `failed to stream` errors.

This was my upgrade process in short:

1. Create a 7.3.2 web/worker AMI/image
2. Set worker Auto Scaling Group to 0 instances
3. Relaunch web instance with new AMI (7.3.2) and wait for dbmigrate to finish.
4. Set worker Auto Scaling Group back to 5 instances with new AMI (7.3.2)

The infra setup/architecture is always the same when I upgrade.
Previous upgrades from 5.x to 6.x and from 6.x to newer 6.x version went perfectly fine this way.

The `failed to stream` issue started when i upgraded to 7.3.2. Same issue with version 7.2.0. I did not test 7.3.0 and 7.3.1.

I needed to rollback to 6.4.0 because devs that are using the ci/cd system were unable to work.

**Reply by evanchaoli** [reply]
Without reproducing the problem, it's hard to tell what was the problem on your deployment. I have a cluster that runs with p2p streaming for long time. If you want to upgrade to 7.x, you may turn off p2p streaming by setting `CONCOURSE_ENABLE_P2P_VOLUME_STREAMING=false`.

**Reply by skabba** [reply]
I understand. Soon I will try to reproduce the problem on a testing environment.

Right after the upgrade to 7.3.2 I did not had the `CONCOURSE_ENABLE_P2P_VOLUME_STREAMING=true` enabled.
If i am understanding it correctly this is an opt-in feature, thus it was disabled right after the upgrade.
Later when troubleshooting, I enabled the P2P option to see if it helped. But it only changed this error: `find or create container on worker ...: failed to stream in to volume` into this error: `find or create container on worker ...: failed to stream p2p out from volume`.

I will get back with the results when i have the testing environment up and running.

**Reply by evanchaoli** [reply]
In that case, that should not be a problem of p2p streaming. Actually p2p streaming only bypasses ATC and streams volume directly from one worker to the other, the logic is quite independent.
